crates/codegen/parser/runtime/src/napi_interface/query.rs

```diff
@@ -0,0 +1,92 @@
+// NAPI-exposed functions have to accept owned values
+#![allow(clippy::needless_pass_by_value)]
+// The functions are meant to be definitions for export, so they're not really used
+#![allow(clippy::return_self_not_must_use)]
+
+use napi::{Env, JsObject};
+use napi_derive::napi;
+
+use crate::napi_interface::cursor::Cursor;
+use crate::napi_interface::{RustQuery, RustQueryResultIterator};
+
+#[napi(namespace = "query")]
+pub struct Query(RustQuery);
+
+impl From<RustQuery> for Query {
+    fn from(value: RustQuery) -> Self {
+        Self(value)
+    }
+}
+
+#[napi(namespace = "query")]
+impl Query {
+    #[napi(factory, catch_unwind)]
+    pub fn parse(text: String) -> napi::Result<Query> {
+        RustQuery::parse(text.as_str()).map_or_else(
+            |err| Err(napi::Error::from_reason(err)),
+            |query| Ok(query.into()),
+        )
+    }
+
+    #[napi(getter, catch_unwind)]
+    pub fn text(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+#[napi(namespace = "query")]
+pub struct QueryResultIterator(RustQueryResultIterator);
+
+#[napi(object, namespace = "query")]
+pub struct QueryResult {
+    pub query_number: u32,
+    #[napi(ts_type = "{ [key: string]: cursor.Cursor[] }")]
+    pub bindings: JsObject,
+}
```

Comment:
nit: I suggest to add a strongly-typed constructor `QueryResult::new()` that takes the cursors map, and create the `JsObject` internally, instead of leaving it up to the caller.